cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -85,67 +87,114 @@ message Result {
+
+* Simpler to maintain and stay upto date with any new ip<>identity changes.
+* All data can be cached in-memory, in an event of agent being unavailable, SDP can lookup from in-memory cache
+
+##### Cons
+
+* Need a method for reconciling the cache in case of SDP restarts.
+* SDP will be aware of all the endpoints data, which might not be needed for DNS proxy.
+
+#### Option 2: Read from bpf maps [Recommended]
+
+We will need to read from couple of bpf maps to get the endpoint ID and identity for a given IP address. For ip to endpoint ID, we can read from `cilium_lxc` and for ip to identity, we can read from `cilium_ipcache`.
+
+##### Pros
 
 * Reduced resource utilization in agent since agent doesn't need to process rpc calls in hotpath
 * IP to endpoint ID or identity translation does not rely on agent availability.
+* DNS can be resolved even when agent is down.
 
-#### Cons
+##### Cons
 
 * Low level interactions from components external to cilium agent.
 
+### Q: Why did we choose to read from bpf maps ?
+
+The reason we chose to read from bpf maps is because the DNS proxy should be able to resolve the DNS queries even when agent is down. The bpf maps are the source of truth for cilium agent and are updated in real time. So, reading from bpf maps ensures that the DNS proxy is always in sync with the cilium agent.
+
+### Flag to disable built-in proxy
+
+Adding a flag(`--disable-builtin-dns-proxy`) to cilium agent to disable the built in dns proxy allows standlone dns proxy to run its independent lifecycle.
+
+#### Pros
+
+* Allows cilium agent to be upgraded without breaking the DNS traffic.
+* Provides a freedom to allow any delegated dns proxy to be plugged in.
+* Reduces the resource utilization in cilium agent.
+
+#### Cons
+
+* Higher latency in DNS resolution as the DNS queries are now handled by a separate process.
+
 ### Q: Restoring L7 DNS policy when agent is unavailable
 
-In an event where SDP restarts when agent's gRPC service is unavailable, can / should SDP rely on information stored on the filesystem to restore per endpoint L7 policy ?
+In an event where SDP restarts when agent's gRPC service is unavailable. SDP can read from the filesystem and restore the policy. The state is stored in `ep_config.json` file. So the flow would be first trying to connect to agent and if it fails, read from the file.
 
+### Scenarios and Expected Behaviour
 
-## Future Milestones
+In an ideal scenario, both cilium agent and SDP should be interacting with each other through a biredirectional grpc stream. However, in an event where either of them restarts/upgrade/donwgrade, what should be the expected behavior ?
 
-* Plumb toFQDN policy for new names when agent is unavailable
+#### Cilium Agent is down, SDP is running
+
+* SDP should be able to serve DNS queries based on the existing bpf maps and in memory DNS rules. The datapath of already configured policies will work. Any new mappings will not be updated until cilium agent is back up.
```

Comment:
Nit: can we paraphrase "able to serve DNS queries" with being able to proxy requests and enforce L7 DNS policy to avoid any confusion for readers.